cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -228,51 +233,87 @@ private Table getMetaTable(DatasetContext context) throws IOException, DatasetMa
    * Get all artifacts that match artifacts in the given ranges.
    *
    * @param range the range to match artifacts in
+   * @param limit the limit number of the result, -1 means return all results
+   * @param order the order of the result, null if unsorted
    * @return an unmodifiable list of all artifacts that match the given ranges. If none exist, an empty list
    *         is returned
    */
-  public List<ArtifactDetail> getArtifacts(final ArtifactRange range) {
+  public List<ArtifactDetail> getArtifacts(final ArtifactRange range, final int limit,
+                                           @Nullable final SortInfo.SortOrder order) {
     try {
       return Transactions.execute(transactional, new TxCallable<List<ArtifactDetail>>() {
         @Override
         public List<ArtifactDetail> call(DatasetContext context) throws Exception {
-          return getArtifacts(getMetaTable(context), range);
+          return getArtifacts(getMetaTable(context), range, limit, order);
         }
       });
     } catch (TransactionFailureException e) {
       throw Transactions.propagate(e);
     }
   }
 
-  private List<ArtifactDetail> getArtifacts(Table metaTable, ArtifactRange range) {
+  private List<ArtifactDetail> getArtifacts(
+    Table metaTable, ArtifactRange range, int limit,
+    @Nullable SortInfo.SortOrder order) throws IOException, ArtifactRangeNotFoundException, ArtifactNotFoundException {
     List<ArtifactDetail> artifacts = Lists.newArrayList();
     ArtifactKey artifactKey = new ArtifactKey(range.getNamespace(), range.getName());
 
     Row row = metaTable.get(artifactKey.getRowKey());
+    if (order == null) {
```

Comment:
its better to avoid nulls if possible. For example, if somebody tries to use then enum in a switch statement, a NPE will be thrown. Let's have an 'unordered' value instead of letting it be null, with the http handler responsible for passing the unordered value if no order is given.